$\large{\textbf{Answer 3.1 Formulation:}}$

$min \ 45x_1 + 48x_2 + 55x_3+ 65x_4 + 54x_5 + 37.5x_6 + 35x_7 + 52.5x_8 + 47x_9 + 28x_{10} + 82x_{11} + 102x_{12} + 150x_{13} + 105x_{14} + 95x_{15} + 84x_{16} + 88x_{17} + 125x_{18} + 70x_{19} + 64x_{20}$

$\text{s.t.} \sum x_i = 425 \\ 0.2x_1 + 0.25x_2 + 0.1x_3 + 0.3x_4 + 0.125x_5 + 0.35x_6 + 0.475x_7 + 0.1x_8 + 0.2x_9 + 0.5x_{10} + 0.2x_{11} + 0.25x_{12} + 0.1x_{13} + 0.3x_{14} + 0.125x_{15} + 0.35x_{16} + 0.475x_{17} + 0.1x_{18} + 0.2x_{19} + 0.5x_{20} = 116.875 \\ 0.75x_1 + 0.65x_2 + 0.7x_3 + 0.6x_4 + 0.85x_5 + 0.525x_6 + 0.425x_7 + 0.4x_8 + 0.35x_9 + 0.4x_{10} + 0.75x_{11} + 0.65x_{12} + 0.7x_{13} + 0.6x_{14} + 0.85x_{15} + 0.525x_{16} + 0.1x_{17} + 0.4x_{18} + 0.35x_{19} + 0.4x_{20} = 244.375 \\ 0.05x_1 + 0.1x_2 + 0.2x_3 + 0.1x_4 + 0.025x_5 + 0.125x_6 + 0.1x_7 + 0.5x_8 + 0.45x_9 + 0.1x_{10} + 0.05x_{11} + 0.1x_{12} + 0.2x_{13} + 0.1x_{14} + 0.025x_{15} + 0.125x_{16} + 0.1x_{17} + 0.5x_{18} + 0.45x_{19} + 0.1x_{20} = 63.75$ 

$l = \{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0\} \textbf{(set of lower bounds)} \\ u = \{15,16,9,17.5,12,18.5,2.5,0.5,9.5,8,+∞,+∞,+∞,+∞,+∞,+∞,+∞,+∞,+∞,+∞\}  \textbf{(set of upper bounds)}$

$\text{where } x_i, 1\leq i \leq 10 \text{ denotes the quantity of } \text{ A-i used from the in house stock and } x_i, 11\leq i \leq 20 \text{ denotes the quantity of } \text{A-i purchased from the market.}$

$\large{\textbf{Answer 3.2 Solution to above problem:}}$

In [41]:
!pip install -q pyomo

In [42]:
from pyomo.environ import *

In [43]:
import numpy as np

In [44]:
# create a model
model = ConcreteModel()

In [45]:
#Setting the values of m and n
M = 4
N = 20

In [46]:
obj_coef_c = np.array([45,48,55,65,54,37.5,35,52.5,47,28,82,102,150,105,95,84,88,125,70,64])

In [47]:
constr_coef_A = np.array([[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
                          [0.2,0.25,0.1,0.3,0.125,0.35,0.475,0.1,0.2,0.5,0.2,0.25,0.1,0.3,0.125,0.35,0.475,0.1,0.2,0.5],
                          [0.75,0.65,0.7,0.6,0.85,0.525,0.425,0.4,0.35,0.4,0.75,0.65,0.7,0.6,0.85,0.525,0.425,0.4,0.35,0.4],
                          [0.05,0.1,0.2,0.1,0.025,0.125,0.1,0.5,0.45,0.1,0.05,0.1,0.2,0.1,0.025,0.125,0.1,0.5,0.45,0.1]])

In [48]:
#Creating the array for RHS of constraints
constr_rhs_b = np.array([425,116.875,244.375,63.75]) 

In [49]:
#Setting the lower and upper bounds
lower_bound = np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0])
upper_bound = np.array([15,16,9,17.5,12,18.5,2.5,0.5,9.5,8, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf])

In [50]:
col_indices = np.arange(N)
row_indices = np.arange(M)

In [51]:
#declare the decision variables in the model
model.x = Var(col_indices)

In [52]:
# create a ConstraintList to hold multiple constraints
model.constraints = ConstraintList()

In [53]:
#Adding the constraints to the constraint list
for i in row_indices:
  model.constraints.add(sum(constr_coef_A[i][j]*model.x[j] for j in col_indices) == constr_rhs_b[i])

In [54]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [55]:
# add the model objective
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [56]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}

1 Var Declarations
    x : Size=20, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  15.0 : False :  True :  Reals
          1 :     0 :  None :  16.0 : False :  True :  Reals
          2 :     0 :  None :   9.0 : False :  True :  Reals
          3 :     0 :  None :  17.5 : False :  True :  Reals
          4 :     0 :  None :  12.0 : False :  True :  Reals
          5 :     0 :  None :  18.5 : False :  True :  Reals
          6 :     0 :  None :   2.5 : False :  True :  Reals
          7 :     0 :  None :   0.5 : False :  True :  Reals
          8 :  

In [57]:
!apt-get install -y -qq glpk-utils

In [58]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 28747.1875
  Upper bound: 28747.1875
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 21
  Number of nonzeros: 81
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01146554946899414
# ----------------------------------------------------------
#   Solution Information
# -------------------------

In [59]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  28747.1875

Decision Variables
x[ 0 ] =  15.0
x[ 1 ] =  16.0
x[ 2 ] =  9.0
x[ 3 ] =  17.5
x[ 4 ] =  12.0
x[ 5 ] =  18.5
x[ 6 ] =  2.5
x[ 7 ] =  0.5
x[ 8 ] =  9.5
x[ 9 ] =  8.0
x[ 10 ] =  158.078125
x[ 11 ] =  0.0
x[ 12 ] =  0.0
x[ 13 ] =  0.0
x[ 14 ] =  0.0
x[ 15 ] =  0.0
x[ 16 ] =  0.0
x[ 17 ] =  0.0
x[ 18 ] =  74.046875
x[ 19 ] =  84.375

Constraints
constraints : Size=4
    Key : Lower   : Body    : Upper
      1 :   425.0 :   425.0 :   425.0
      2 : 116.875 : 116.875 : 116.875
      3 : 244.375 : 244.375 : 244.375
      4 :   63.75 :   63.75 :   63.75


$\large{\textbf{Answer 3.3:}} \\ \textbf{Objective function value = Min. Cost = }28747.1875 \\ \textbf{Optimal Values of decision variables:} \\ x_1 = 15 \ , \ x_2 = 16 \ , \ x_3 = 9 \ , \ x_4 = 17.5 \ , \ x_5 = 12 , \\ x_6 = 18.5 \ , \ x_7 = 2.5 \ , \ x_8 = 0.5 \ , \ x_9 = 9.5 \ , \  x_{10} = 8 , \\ x_{11} = 158.078125 \ , \  x_{12} = 0 \ , \  x_{13} = 0 \ , \  x_{14} = 0 \ , \  x_{15} = 0 , \\  x_{16} = 0 \ , \ _{17} = 0 \ , \  x_{18} = 0 \ , \  x_{19} = 74.046875 \ , \  x_{20} = 84.375$

$\large{\textbf{Total quantities of alloys used:}} \\ A-1 = x_1 + x_{11} = 15 + 158.078125 = 173.078125 \\ A-2 = x_2 + x_{12} = 16 + 0 = 16 \\ A-3 = x_3 + x_{13} = 9 + 0 = 9 \\ A-4 = x_4 + x_{14} = 17.5 + 0 = 17.5 \\ A-5 = x_5 + x_{15} = 12 + 0 = 12 \\ A-6 = x_6 + x_{16} = 18.5 + 0 = 18.5 \\ A-7 = x_7 + x_{17} = 2.5 + 0 = 2.5 \\ A-8 = x_8 + x_{18} = 0.5 + 0 = 0.5 \\ A-9 = x_9 + x_{19} = 9.5 + 74.046875 = 83.546875 \\ A-10 = x_{10} + x_{20} = 8 + 84.375 = 92.375$

$\large{\textbf{Answer 3.4 :}} \\ \text{Only three alloys are required to be bought from the market, rest are only used from the in-house stock.} \\ \text{The alloys ranked in the decreasing order of quantities bought are: } \\ 1. \ A-1 \ = 158.078125 \\ 2. \ A-10 \ = 84.375 \\ 3. \ A-9  \ =74.046875 \\ \text{Rest are not purchased from the market so the quantities bought are zero. Hence they are equally ranked at the bottom.}$ 


$\large{\textbf{Answer 3.5 : }} \\ \text{All the alloys are completely utilised as the in-house cost is comparitively very low from the purchasing cost,} \\ \text{ so the alloys are first used from the stock then they are bought from the market. }$

$\large{\textbf{Answer 3.6 :}} \\ \text{The alloys which are not acquired from the market are : } \\ 1. \ A-2 \\ 2. \ A-3 \\ 3. \ A-4 \\ 4. \ A-5 \\ 5. \ A-6 \\ 6. \ A-7 \\ 7. \ A-8 \\ \text{The purchasing cost of the above stated alloys is higher than the purchasing cost of the rest of the three alloys,} \\ \text{hence the alloys with the lower costs are first used and they satisfied the demand of the constraints so alloys with higher costs are not used.}$